src/equations/serre_green_naghdi_1d.jl

```diff
@@ -97,59 +181,191 @@ function create_cache(mesh,
-                 D, Dmat_minus, factorization)
+                 D1, D1mat_minus, factorization)
     else
-        if D isa FourierDerivativeOperator
-            Dmat = Matrix(D)
+        if D1 isa FourierDerivativeOperator
+            D1mat = Matrix(D1)
 
-            cache = (; h_x, v_x, h2_x, hv_x, v2_x,
+            cache = (; h, b, h_x, v_x, h2_x, hv_x, v2_x,
                      h2_v_vx_x, h_vx_x, p_x, tmp,
                      M_h, M_h3_3,
-                     D, Dmat)
+                     D1, D1mat)
         else
-            Dmat = sparse(D)
+            D1mat = sparse(D1)
 
+            @. M_h = h
+            scale_by_mass_matrix!(M_h, D1)
+            @. M_h3_3 = (1 / 3) * h^3
+            scale_by_mass_matrix!(M_h3_3, D1)
             # Floating point errors accumulate a bit and the system matrix
             # is not necessarily perfectly symmetric but only up to
             # round-off errors. We wrap it here to avoid issues with the
             # factorization.
-            @. M_h = h
-            scale_by_mass_matrix!(M_h, D)
-            @. M_h3_3 = (1 / 3) * h^3
-            scale_by_mass_matrix!(M_h3_3, D)
             system_matrix = Symmetric(Diagonal(M_h)
                                       +
-                                      Dmat' * Diagonal(M_h3_3) * Dmat)
+                                      D1mat' * Diagonal(M_h3_3) * D1mat)
+
             factorization = cholesky(system_matrix)
 
-            cache = (; h_x, v_x, h2_x, hv_x, v2_x,
+            cache = (; h, b, h_x, v_x, h2_x, hv_x, v2_x,
                      h2_v_vx_x, h_vx_x, p_x, tmp,
                      M_h, M_h3_3,
-                     D, Dmat, factorization)
+                     D1, D1mat, factorization)
+        end
+    end
+
+    return cache
+end
+
+# variable bathymetry
+function create_cache(mesh,
+                      equations::SerreGreenNaghdiEquations1D,
+                      solver,
+                      initial_condition,
+                      ::BoundaryConditionPeriodic,
+                      RealT, uEltype)
+    D1 = solver.D1
+
+    # create temporary storage
+    h = ones(RealT, nnodes(mesh))
+    h_x = zero(h)
+    b = zero(h)
+    b_x = zero(h)
+    v_x = zero(h)
+    h_hpb_x = zero(h)
+    hv_x = zero(h)
+    v2_x = zero(h)
+    h2_v_vx_x = zero(h)
+    h_vx_x = zero(h)
+    p_h = zero(h)
+    p_x = zero(h)
+    tmp = zero(h)
+    M_h_p_h_bx2 = zero(h)
+    M_h3_3 = zero(h)
+    M_h2_bx = zero(h)
+
+    # b_x appears in the system matrix, so it must not be zero to get the
+    # correct pattern of the factorization. Just setting it to unity does
+    # also not work since some terms cancel. Thus, b_x needs to be
+    # "variable enough" to get the correct pattern of non-zero entries in
+    # the system matrix and the factorization.
+    # TODO: This is a hack and should be improved. It would be ideal if we
+    #       had access to the initial condition and the exact value of the
+    #       bathymetry here.
```

Comment:
Wait...don't we _have_ access to the initial condition, cf. line 245?

Reply:
But if we evaluated it here, we would evaluate it twice. Moreover, for the case of time-dependent bathymetry that you have in mind, the initial bathymetry may still be constant so that we need a hack like this.

Reply:
I see. In that case I also don't know of any good solution.

Reply:
We just keep it as it is for this PR?